### PR TITLE
New rule: import-blacklist

### PR DESCRIPTION
**Pull request by SimonSchick**
#### PR checklist

- [ ] Addresses an existing issue: #0000
- [x] New feature, bugfix, or enhancement
  - [x] Includes tests
- [x] Documentation update

#### What changes did you make?

Cont. of https://github.com/palantir/tslint/pull/1549

I was given permission by the author to continue his work:
I added tests.
Fixed problems detected by tests (somewhat of a refactor too).
Added license.

Also due to a previous change master broke, I fixed this in my PR to make it build for now.

#### Is there anything you'd like reviewers to focus on?

Perhaps a better error message? I'm not good with words.

### CODE REVIEW

**Comment by SimonSchick** [comment]
Don't want to spam you @use-strict but please sign the palantir CLA in your branch to get this rolling, a review is also welcome 😄

**Comment by SimonSchick** [comment]
It seems that the doc generator also included some things from other rules that apparently had no doc generated yet? Is this expected? Please advise.

**Comment by use-strict** [comment]
@SimonSchick , done. Great job, thanks!

**Comment by SimonSchick** [comment]
Noticed conflict, will merge/rebase after I got approval.

File: src/rules/importBlacklistRule.ts [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+        optionsDescription: "A list of blacklisted modules.",
+        options: {
+            type: "array",
+            items: {
+                type: "string",
+            },
+            minLength: 1,
+        },
+        optionExamples: ["true", '[true, "rxjs", "lodash"]'],
+        type: "typescript",
```
Comment (adidahiya) [inline src/rules/importBlacklistRule.ts]: I don't think this is typescript-only -- I'd throw it in the "functionality" bucket (kind of a catch-all)

File: src/rules/importBlacklistRule.ts [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+        typescriptOnly: false,
+    };
+
+    public static FAILURE_STRING = "This import is not allowed.";
```
Comment (adidahiya) [inline src/rules/importBlacklistRule.ts]: Suggestion: "This import is blacklisted, consider importing a submodule instead"

Minor: we're a little inconsistent about this elsewhere, but I'd prefer to omit the trailing period in this failure string.

File: src/rules/importBlacklistRule.ts [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+        );
+    }
+}
+
+class NoRequireFullLibraryWalker extends Lint.RuleWalker {
+    constructor (sourceFile: ts.SourceFile, options: Lint.IOptions, private blackList: string[]) {
+        super(sourceFile, options);
+    }
+
+    protected isListed(text: string): boolean {
```
Comment (adidahiya) [inline src/rules/importBlacklistRule.ts]: why `protected`?

File: src/rules/importBlacklistRule.ts [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+    public apply(sourceFile: ts.SourceFile): Lint.RuleFailure[] {
+        const options = this.getOptions();
+        return this.applyWithWalker(
+            new NoRequireFullLibraryWalker(sourceFile, options, options.ruleArguments),
+        );
+    }
+}
+
+class NoRequireFullLibraryWalker extends Lint.RuleWalker {
+    constructor (sourceFile: ts.SourceFile, options: Lint.IOptions, private blackList: string[]) {
```
Comment (adidahiya) [inline src/rules/importBlacklistRule.ts]: minor: can you remove the `private` parameter property syntax and do this the old fashioned way, `this.blacklist = blacklist`? I find that more readable. (note the lack of camelCase, just "blacklist", not "blackList".)

File: src/rules/importBlacklistRule.ts [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+    }
+
+    public visitImportDeclaration(node: ts.ImportDeclaration) {
+        if (this.isListed(node.moduleSpecifier.getText())) {
+            this.failure(node.moduleSpecifier);
+        }
+        super.visitImportDeclaration(node);
+    }
+
+    private failure (node: ts.Node): void {
```
Comment (adidahiya) [inline src/rules/importBlacklistRule.ts]: question: is there any more specific common type for the `node` parameter here that we can declare?

minor: can you please call this method `reportFailure` to be more verbose / readable

File: src/rules/importBlacklistRule.ts (hunk 3 above)
Comment (adidahiya) [inline src/rules/importBlacklistRule.ts]: also I think a nicer method signature would be `private isModuleBlacklisted(moduleName: string): boolean`

File: src/rules/importBlacklistRule.ts (hunk 1 above)
Comment (SimonSchick) [inline src/rules/importBlacklistRule.ts]: Oh yea, that is a copy paste error, sorry 😛

File: src/rules/importBlacklistRule.ts (hunk 2 above)
Comment (SimonSchick) [inline src/rules/importBlacklistRule.ts]: sure thing

Comment (SimonSchick) [inline src/rules/importBlacklistRule.ts]: I think I gonna go for a slightly more imperative and concise message, feel free to complain tho 😛

**Comment by SimonSchick** [comment]
Happy? 😄

**Comment by adidahiya** [comment]
looks great, thanks @SimonSchick!

**Pull request merged**